### ISSUE TITLE
Generating data from a derived class with MemberData

### ISSUE THREAD

**Issue by jamesqo** [xunit/xunit#993]
Hi, I'm wondering if it's possible for a derived class to affect the `MemberData` methods of a base class. Allow me to explain. I'm writing tests for a generic collection I've written that look like this:

```cs
public abstract class MyCollectionTests<T>
{
    [Theory]
    [MemberData(nameof(EnumerableData)]
    public void ToArray(IEnumerable<T> seed)
    {
    }

    public static TheoryData<IEnumerable<T>> EnumerableData()
    {
        get { ... }
    }
}

public class MyCollectionTestsInt32 : MyCollectionTests<int> { }
public class MyCollectionTestsString : MyCollectionTests<string> { }
```

Now when you get to `EnumerableData`, you're stuck at this point because the only Ts you can generate are `default(T)`s. But you don't want to test just default values, you might want to throw in `"foo"`s and `"bar"`s and `0xf00`. The instinctive reaction for me was to put in an abstract method in the derived classes:

```cs
public abstract class MyCollectionTests<T>
{
    protected abstract T GenerateItem();

    private IEnumerable<T> GenerateItems(int count)
    {
        for (int i = 0; i < count; i++)
        {
            yield return GenerateItem();
        }
    }
}

public class MyCollectionTestsInt32 : MyCollectionTests<int>
{
    protected override int GenerateItem() => ...
}
```

Unfortunately I realized that all MemberDatas had to be static, which made be try giving the derived classes a static constructor and having them set a static property in the base class:

```cs
public abstract class MyCollectionTests<T>
{
    protected static Func<T> GenerateItem { get; set; }

    public static TheoryData<IEnumerable<T>> EnumerableData()
    {
        get
        {
            return new TheoryData<IEnumerable<T>>
            {
                // force enumeration of sequence while MemberData is being evaluated
                { GenerateItems(someCount).ToList() }
            };
        }
    }

    private static IEnumerable<T> GenerateItems(int count)
    {
        Debug.Assert(GenerateItem != null);

        for (int i = 0; i < count; i++)
        {
            yield return GenerateItem();
        }
    }
}

public class MyCollectionTestsInt32 : MyCollectionTests<int>
{
    private static int s_state;

    static MyCollectionTestsInt32()
    {
        GenerateItem = () => s_state++;
    }
}
```

But this does not work either, since sometimes I am hitting the `GenerateItem != null` assert while running, indicating that somehow `GenerateItems` must have run before the static constructors were executed.

So my question is: is this expected behavior? Also, what's the correct way to get the derived classes to feed information into the generic base class so it can use it in `MemberData`?

Thanks.

**Comment by hughbe** [comment]
@jamesqo take a look at #967, would this help with your problem/

**Comment by bradwilson** [comment]
Fixed by #967.
